src/smallstring.jl

```diff
@@ -1,66 +1,86 @@
+export convert,
+       push,
+       setindex
 
 const IntChar = UInt8
 const IntSmallString = UInt64
 const maxTagLength = 8
 const SmallStringStorage = SVector{maxTagLength,IntChar}
 const MSmallStringStorage = MVector{maxTagLength,IntChar} # Mutable SmallString storage
 
+function initSmallStringStore()
+  return SmallStringStorage(ntuple(_->IntChar(0),Val(length(SmallStringStorage))))
+end
+
 struct SmallString
   data::SmallStringStorage
-  length::Int
-  SmallString(sv::SmallStringStorage,l::Int) = new(sv,l)
+
+  SmallString(sv::SmallStringStorage) = new(sv)
+
   function SmallString()
-    sv = SmallStringStorage(ntuple(_ -> IntChar(0),Val(length(SmallStringStorage))))
-    return new(sv, 0)
+    sv = initSmallStringStore()
+    return new(sv)
   end
-end
 
-function SmallString(sv::SmallStringStorage)
-  sv[1] == IntChar(0) && return SmallString(sv,0)
-  len = 1
-  while len < length(SmallStringStorage) && @inbounds sv[len+1] ≠ IntChar(0)
-    len += 1
+  function SmallString(str::String)
+    sv = initSmallStringStore()
+    n = 1
+    while n <= length(str) && n <= maxTagLength
+      sv = setindex(sv,IntChar(str[n]),n)
+      n += 1
+    end
+    return new(sv)
   end
-  SmallString(sv,len)
 end
 
-Base.length(s::SmallString) = s.length
+SmallString(s::SmallString) = SmallString(s.data)
+
+#Base.length(s::SmallString) = s.length
+
 Base.getindex(s::SmallString,n::Integer) = getindex(s.data,n)
+
 function Base.setindex(s::SmallString,val,n::Integer)
-  len = length(s)
-  if n > len
-    len = n
-  end
-  return SmallString(setindex(s.data,val,n),len)
+  return SmallString(setindex(s.data,val,n))
 end
 
 isNull(s::SmallString) = @inbounds s[1] == IntChar(0)
 
 function StaticArrays.push(s::SmallString,val)
-  len = length(s)
-  return SmallString(setindex(s.data,val,len+1),len+1)
+  newlen = 1
+  while newlen <= maxTagLength && s[newlen] != IntChar(0)
+    newlen += 1
+  end
```

Comment:
The reason I put the length into `SmallString` (as I am sure you know) was to avoid having to do checks like this. Was the reason to remove it because we want `SmallString` to be `isbits`? It seems like we can still have that:

```julia
struct SmallString
  data::SVector{8,UInt8}
  length::Int
end

isbitstype(SmallString)   # I get true
```

Did that not work with the previous definition? I think structs of isbits types are still isbits.

Reply:
Yes, I did jump the gun here. I thought the layout on master wasn't isbits but I was wrong about that.

There's a tradeoff here between memory and computation to be sure. Including the length would make the memory footprint exactly twice as large (since the length is also a 64 bit integer). But it would still meet the key design goal of not allocating.

I had forgotten, but this was the reason I actually capped the length of small strings at 7 in the C++ version. Then the last character could always be guaranteed to be equal to IntChar(0), through the constructor's behavior, and then it is safe for loops to only check equivalence to IntChar(0) and not have to also compare to the max length.

Would you be ok to cap the length at 7 (still keeping the storage at 8 chars like it is now)? Then we could have both 64 bits of memory and speed both! The only downside is having to keep tags shorter.

Reply:
Yes I guess that makes the most sense. I wasn't thinking about memory, but an ITensor could end up with quite a lot of tags. Too bad we can't have the tag "Electron".

Reply:
That is a shame huh. Let's think about this some more. I was thinking the other day that one might want to use `Electron` operators for spinful fermions which aren't electrons (e.g. protons are one key example). So maybe there's a better name anyway. But SFermion is also 8 characters! Is there another good name? Elec? SpinF? Hm...hard to think of one.

(Sidebar: If we wanted to get really low-level with this, I just read that ASCII characters only use their first 7 bits anyway. So in principle one can pack the bits to the front and get some at the end back even with 64 bits. There's even an experimental Julia code that does something like this! https://github.com/tpapp/PackedASCIITypes.jl  But it's probably overkill. And it causes extra cpu steps to do things like retrieving a single character.)

Reply:
In the C++ version we use "Elec", I can't really think of anything better either. Protons are a good example, maybe with the new tag-based system we can think of an easy way to let people use multiple names to call the same operators? (You could always just overload `op` and `state` on a ProtonType to call the ElectronType functions). It may be nice to define both "Elec" and "Proton" indices even if they use the same operators.

I also just realized the length doesn't need to by an Int64, just a UInt8, since of course it is always positive and not larger than UInt8. So maybe it is not too bad to store the length?

That package seems quite close to what we want, too bad accessing elements is not very good, we don't want to convert to strings all the time.

Reply:
One possibility is we keep it at 8, don't put the length back in, and then just allow the extra computation of comparing to the maximum length to happen. It's not perfectly ideal, but I think in practice it will be only a very little bit extra. I'm willing to bet that comparisons of SmallStrings are really still how they mostly get used, and those remain fast.

Doing a survey of the code as it stands, the double comparison (to both IntChar(0) and smallLength) only happens in `push`, `String(SmallString)`, `show` and `isint`. However, push and the conversion to String aren't used anywhere in the code right now! The show function for small strings doesn't happen in any speed-critical loops. Quite probably `isint` never gets called in such loops either.

Reply:
Yeah that seems fine. I guess the key decision we want to make now is how long the tag names should be, because that would be an annoying breaking change to make later on if we set them at 8 and then switched it back down to 7. But with 8, it looks like we have two different design options that we can test out for efficiency, so it doesn't seem to lock us into a single design. It is too bad `isint` needs to get called every time a tag is made, perhaps we can even loosen that restriction or make that a debug check.

`push` was there when I was A/B testing whether it was better to fill an `MTagSetStorage` when creating a TagSet or replace with a new `TagSetStorage` as tags got added, I am fine with removing it. The first version seemed to be faster, but it is something to keep in mind if we want to revisit optimizing the TagSet code later on.

Reply:
I think keeping it at 8 is the right call so we can use good names.

Thanks for that info about push. Regarding isint, yes we could perhaps turn it off, especially now that primes are passed separately in a tuple (though it probably remains a good idea to disallow integer tags).

There may be one nice way to speed up isint I thought of today: just fully unroll the loop and write a function of 10 lines like this:
```Julia
is_int = false
is_int = (s[1] != IntChar(0) && isdigit(Char(s[1])))
is_int = (s[1] != IntChar(0) && isdigit(Char(s[1]))) 
is_int = (s[1] != IntChar(0) && isdigit(Char(s[1]))) 
```

Reply:
Interesting idea. We can probably write a macro for that. I guess the over all conclusion though is that it is probably fast enough as is, so we can go with 8 characters and play with optimizations later if needed.

Reply:
Ok yes I'm good with that conclusion too. Based on that then, do you think this is ok to pull at this point?